File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/domain/role/Role.java
package br.com.deltaglobalbank.identity.domain.role;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

public class Role {

    private final UUID id;
    private final RoleCode code;
    private final UUID moduleId;
    private final String description;
    private final String label;
    private final boolean active;
    private final Instant createdAt;

    public Role(
        UUID id,
        RoleCode code,
        UUID moduleId,
        String description,
        String label,
        boolean active,
        Instant createdAt
    ) {
        this.id = id;
        this.code = code;
        this.moduleId = moduleId;
        this.description = description;
        this.label = label;
        this.active = active;
        this.createdAt = createdAt;
    }

    public Role activate() {
        return new Role(id, code, moduleId, description, label, true, createdAt);
    }

    public Role deactivate() {
        return new Role(id, code, moduleId, description, label, false, createdAt);
    }

    public UUID getId() {
        return id;
    }

    public RoleCode getCode() {
        return code;
    }

    public UUID getModuleId() {
        return moduleId;
    }

    public String getDescription() {
        return description;
    }

    public String getLabel() {
        return label;
    }

    public boolean isActive() {
        return active;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public RoleSnapshot snapshot() {
        return new RoleSnapshot(
            id,
            code,
            moduleId,
            description,
            label,
            active,
            createdAt
        );
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof Role role && id.equals(role.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Role(id=" + id + ", code=" + code + ")";
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/domain/role/Role.kt
package br.com.deltaglobalbank.identity.domain.role

import java.time.Instant
import java.util.UUID

class Role(
    val id: UUID,
    val code: RoleCode,
    val moduleId: UUID?,
    val description: String?,
    val createdAt: Instant
) {

    fun snapshot(): RoleSnapshot = RoleSnapshot(
        id = id,
        code = code,
        moduleId = moduleId,
        description = description,
        createdAt = createdAt
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Role) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String = "Role(id=$id, code=$code)"
}

data class RoleSnapshot(
    val id: UUID,
    val code: RoleCode,
    val moduleId: UUID?,
    val description: String?,
    val createdAt: Instant
)
